---
title: "dev_forum"
author: "Justin Finkle"
date: "8/26/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(org.Hs.eg.db)
library(plotly)
library(chipVis)
```

```{r Load Annotation Data}
txdb <- TxDb.Hsapiens.BioMart.igis
tx_data <- load_tx_data(txdb)
symbol_table <- match_tx_to_gene(txdb, 'org.Hs.eg')
```

```{r Load Data}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000

T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)
```

# Plotting multiple genes for several conditions
```{r}
# set the gene list
entrez_id <- c(1509, 183, 2161, 2353, 3265, 3934, 4057, 5016, 5020, 9429, 526, 9687)
genes <- (symbol_table %>% filter(entrez %in% entrez_id))$symbol
genes <- c('IL20RA', 'CFAP20', 'PACSIN1', 'ST6GALNAC1', 'GREB1',
           'ZNF579', 'PRKAB2')
# Select the indices of the chip-seq experiments
# c_idx stands for conditions index
c_idx <- sapply(T47D[c(1,2,4,5)], function(x){x[[1]]})

# Set how far upstream and downstream  you want to view from the TSS
extension <- 20000
```

```{r Load Expression}
rpkmEset <- readRDS("/gne/research/workspace/schepa/Carroll_Data/GSE68358.rpkmEset.with_fdata.Rds")
T47D_eset <- rpkmEset[,grep("T47D",pData(rpkmEset)$celltype)]
treat_groups = stringr::str_trim(sapply(stringr::str_split(pData(T47D_eset)$treatment,":"), 
                      function(x) x[[2]])) %>% gsub("+","_",., fixed= TRUE) %>% 
  gsub("_3hr","",.)

library(limma)
library(edgeR)

dge <- DGEList(counts = exprs(T47D_eset)) %>% calcNormFactors()
design <- model.matrix(~ 0 +treat_groups)
colnames(design) = gsub("treat_groups", "", colnames(design))
v <- voom(dge, design=design)

fit <- lmFit(v, design)
contrasts <- makeContrasts(E2_Progesterone-E2, 
                           E2_R5020-E2, 
                           E2_R5020-E2_Progesterone, levels = design)
fit2 <- contrasts.fit(fit, contrasts) %>% eBayes()
results_Progesterone <- topTable(fit2, coef = 1, adjust = "BH")
results_R5020 <- topTable(fit2, coef = 2, adjust = "BH")

results <- decideTests(fit2, p.value = 0.01, lfc = 1)
sig <- which(apply(results,1, function(x) sum(x!=0) > 0))
sig_genes <- fData(T47D_eset)$symbol[sig]
```

```{r Expression Bar Plots}
gene <- genes[[7]]
expression <- v$E[which(fData(T47D_eset)$symbol==gene),]
names(expression) <- treat_groups
e2 <- expression[names(expression)=='E2']
pr <- expression[names(expression)=='E2_Progesterone']
plot_ly(y = e2, type = "box", name = "E2", pointpos=0, boxpoints='all') %>% 
  add_trace(y = pr, type = "box", name = 'PR', boxpoints='all', pointpos=0) %>%
  layout(yaxis=list(title='Expression'))
```

```{r Multiple Genes, echo=FALSE}
lib_size <- get_scaling_factor(fi[c_idx,])[2,]/1000000
conditions <- names(c_idx)
p_list <- list()
a_list <- list()
r_list <- list()
score_max <- 0
for(gene in genes){
  sl <- ifelse(gene == genes[1], TRUE, FALSE)
  p <- plot_ly(type = 'scatter', showlegend=sl, 
               name = gene,
               colors = rev(RColorBrewer::brewer.pal(4, 'Dark2')))
  x <- which(symbol_table$symbol == gene)
  
  # Get the range of the gene
  cur_tss <- ifelse(as.character(symbol_table$strand[x]) == "-",
                    symbol_table$end[x],
                    symbol_table$start[x])
  cur_range <- get_view_range(chr = symbol_table$chr[x],
                              start = cur_tss - extension,
                              end = cur_tss + extension,
                              strand = as.character(symbol_table$strand[x]))
  # Get the coverage in the range
  cur_cvg <- make_coverage_tracks(fi[c_idx, "File.bw"], 
                            target_range = cur_range, 
                            sample_names = conditions, 
                            scaling_factors = lib_size)
  # Get max score for scaling y_axes
  score_max <- max(max(as.data.frame(mcols(cur_cvg)), score_max))
  cvg_df <- biovizBase::mold(cur_cvg)
  
  # Get the relative distances
  cvg_df$midpoint <- cvg_df$midpoint - cur_tss
  
  # Add the coverage as traces
  for(cond in conditions){
    p <- p %>% add_trace(x = cvg_df$midpoint, y = cvg_df[[cond]], mode='lines',
                         name = cond, color = cond)
  }
  
  #Save the annotation information
  cur_tx <- get_tx_annotation(range=cur_range, tx_data = tx_data)
  cur_tx$stepping <- -1
  cur_tx <- biovizBase::mold(cur_tx)
  cur_tx$start <- cur_tx$start - cur_tss
  cur_tx$end <- cur_tx$end - cur_tss
  cur_tx$midpoint <- cur_tx$midpoint - cur_tss
  a_list[[gene]] <- cur_tx
  start(cur_range) <- start(cur_range) - cur_tss
  end(cur_range) <- end(cur_range) - cur_tss
  r_list[[gene]] <- cur_range
  
  # Add invisible annotation handles
  h_features <- unique(cur_tx$feature)
  for(h_feature in h_features){
    h <- cur_tx[cur_tx$feature == h_feature, ]
    p <- p %>% add_trace(x=h$midpoint, 
                         y=h$stepping, 
                         opacity=0,
                         text = h$transcript, 
                         name = h_feature,
                         # color = h_feature,
                         # colors = rep('grey', length.out = length(h_features)),
                         hoverinfo="x+name+text",
                         marker = list(color='grey'),
                         showlegend=FALSE)
  }
  p_list[[gene]] <- p
}
sp <- subplot(p_list, nrows=length(p_list), shareX = TRUE)

```

```{r Add annotations, echo=FALSE}
trace_names <- sapply(sp$x$data, function(x){x$name})
trace_axes <- lapply(sp$x$data, function(x) {gsub("y", "yaxis", x$yaxis)})
names(trace_axes) <- trace_names  
sp2 <- sp
for(gene in genes){
  yax <- trace_axes[[gene]]
  sp2$x$layout[[yax]][['title']] <- gene
  sp2$x$layout[[yax]][['zeroline']] <- FALSE
  sp2$x$layout[[yax]][['range']] <- c(-(0.15*score_max), score_max)
  sp2 <- add_tx_shapes(sp2, a_list[[gene]], r_list[[gene]], yax, y_scaling = score_max)
}
sp2$x$layout$showlegend <- TRUE
sp2$x$layout$xaxis$range <- c(-extension, extension)
sp2$x$layout$xaxis$title <- 'Distance to TSS'
sp2$x$layout$xaxis$zeroline <- FALSE
tss_line <- list(x0 = 0, x1 = 0, xref = 'x', y0=0, y1=1, yref='paper', type='line',
                 opacity = 0.1)
sp2$x$layout$shapes[[length(sp2$x$layout$shapes)+1]] <- tss_line

sp2
```

```{r}
# Save the plot
export(sp2, file = '/gne/research/workspace/finklej/intern_poster/mesoscale_tracks_wide.png')
```
