---
title: "Introduction"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(iheatmapr)
library(chipVis)
library(dplyr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
```

## Example


```{r}
bamfiles <- list.files(system.file("extdata", package = "genomationData"),
           pattern = "*.bam$", full.names = TRUE)

samp.file = system.file('extdata/SamplesInfo.txt',package='genomationData')
samp.info = read.table(samp.file, header=TRUE, sep='\t', 
                       stringsAsFactors = FALSE)
names(bamfiles) <- left_join(data_frame(fileName = basename(bamfiles)), samp.info) %>%
  select(sampleName) %>% unlist(use.names = FALSE)


promoter_regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                              upstream = 1000, downstream = 1000)
promoter_regions <- promoter_regions[seqnames(promoter_regions) == "chr21"]

cov_mats <- make_coverage_matrix(bamfiles, promoter_regions, binsize = 20)

ctcf.peaks = genomation::readBroadPeak(system.file("extdata",
                         "wgEncodeBroadHistoneH1hescCtcfStdPk.broadPeak.gz",
                         package = "genomationData"))
ctcf.peaks = ctcf.peaks[seqnames(ctcf.peaks) == "chr21"]
ctcf.peaks = ctcf.peaks[order(-ctcf.peaks$signalValue)]
ctcf.peaks = resize(ctcf.peaks, width = 1, fix = "center")

ctcf_mats <- make_coverage_matrix(bamfiles, ctcf.peaks, 
                                  up = 1000, down = 1000, 
                                  binsize = 25)

coverage_heatmap(ctcf_mats[[1]])

```

```{r}
coverage_heatmap(ctcf_mats)
```



# Genomics Heatmaps

## Expression heatmap

```{r}
data("rpkm_chr21")
```

```{r}
iheatmap(assays(rpkm_chr21)[["rpkm"]])
```

```{r}
iheatmap(rpkm_chr21,"rpkm")
```

```{r}
iheatmap(rpkm_chr21, "rpkm",
         x = colData(rpkm_chr21)$STD_NAME, y = rowData(rpkm_chr21)$SYMBOL, 
         col_annotation = colData(rpkm_chr21)[,c("TYPE","SEX")])
```

There is also an `add_iheatmap` method defined for SummarizedExperiment for adding a heatmap based on a SummarizedExperiment to an existing heatmap.

## Coverage heatmaps

### The data

```{r}
genomation_dir <- system.file("extdata", package = "genomationData")

samp.file <- file.path(genomation_dir,'SamplesInfo.txt')
samp.info <- read.table(samp.file, header=TRUE, sep='\t', 
                       stringsAsFactors = FALSE)
samp.info$fileName <- file.path(genomation_dir, samp.info$fileName)

```

```{r}
ctcf.peaks = genomation::readBroadPeak(system.file("extdata",
                         "wgEncodeBroadHistoneH1hescCtcfStdPk.broadPeak.gz",
                         package = "genomationData"))
ctcf.peaks = ctcf.peaks[seqnames(ctcf.peaks) == "chr21"]
ctcf.peaks = ctcf.peaks[order(-ctcf.peaks$signalValue)]
ctcf.peaks = resize(ctcf.peaks, width = 501, fix = "center")
```

### Making coverage matrices

```{r}
ctcf_mats <- make_coverage_matrix(bamfiles, 
                                  ctcf.peaks, 
                                  up = 1000, down = 1000, 
                                  binsize = 25)

```


### Single coverage

```{r}
coverage_heatmap(ctcf_mats[[1]])
```


### Adding coverage heatmap

### Multiple coverage

### Add multiple coverage heatmaps

### Coverage heatmap at TSS

### Combining expression heatmap with coverage heatmap

# Interactive genome tracks

## View ranges

## Pre-load annotation data

## Single track

## Multiple tracks

## Expression




## Single locus

### View Range


```{r view_range}
view_range <- make_view_range(chr = "chr21", start=45050000, end=45068000)
view_range
```

### Annotation data
We pre-load the annotation information from the TxDB into an easily searched list. This saves time when repeated calls are made to find the genomic ranges and features that are used in plotting coverage information. 
The variable `tx_data` contains several lists that a indexed by transcript
feature (e.g. intron, utr5). It also keeps track of the seqlevelstyle used
in the database.

```{r Load Annotation reference}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
tx_data <- load_tx_data(txdb)
print(names(tx_data))
```


## Track plotter

```{r}
track_plotter <- make_track_plotter(samp.info$sampleName[1:3], tx_data, 
                                    track_names = samp.info$sampleName[1:3] , 
                                    share_y = TRUE)
```

```{r}
track_plotter(ctcf.peaks[1])
```




## Standard plot

```{r Demo_stacking}

```

By default, browserly densely stacks all overlapping transcripts to save visual space. Future versions may allow for the display of "canonical" transcripts. All y axes default to being on the same scale, but this can also be disabled.

## Heatmap of coverage
Browserly automatcially determines the type of plot for showing coverage. When many
samples are displayed, browserly defaults to using a heatmap. This saves vertical space, without losing information.
```{r Demo_hm}
plot_single_locus(view_range, tx_data, cvg)
```


```{r}
tr_fn <- make_browserly_function(bamfiles[c("Ctcf","Znf143","Suz12")], tx_data = tx_data,
                                cvg_scaling = 100)
```


## Multiple locus



## Adding expression information




## Log scale???

# Apps linking heatmaps to genome tracks


```{r}

```





