---
title: "Introduction"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(iheatmapr)
library(chipVis)
library(dplyr)
```

## Example


```{r}
bamfiles <- list.files(system.file("extdata", package = "genomationData"),
           pattern = "*.bam$", full.names = TRUE)

samp.file = system.file('extdata/SamplesInfo.txt',package='genomationData')
samp.info = read.table(samp.file, header=TRUE, sep='\t', 
                       stringsAsFactors = FALSE)
names(bamfiles) <- left_join(data_frame(fileName = basename(bamfiles)), samp.info) %>%
  select(sampleName) %>% unlist(use.names = FALSE)



library(TxDb.Hsapiens.UCSC.hg19.knownGene)
promoter_regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                              upstream = 1000, downstream = 1000)
promoter_regions <- promoter_regions[seqnames(promoter_regions) == "chr21"]

cov_mats <- make_coverage_matrix(bamfiles, promoter_regions, binsize = 20)

ctcf.peaks = genomation::readBroadPeak(system.file("extdata",
                         "wgEncodeBroadHistoneH1hescCtcfStdPk.broadPeak.gz",
                         package = "genomationData"))
ctcf.peaks = ctcf.peaks[seqnames(ctcf.peaks) == "chr21"]
ctcf.peaks = ctcf.peaks[order(-ctcf.peaks$signalValue)]
ctcf.peaks = resize(ctcf.peaks, width = 1, fix = "center")

ctcf_mats <- make_coverage_matrix(bamfiles, ctcf.peaks, up = 1000, down = 1000, 
                                  binsize = 25)

coverage_heatmap(ctcf_mats[[1]])

```

```{r}
coverage_heatmap(ctcf_mats)
```



# Genomics heatmaps

## Coverage heatmaps

### Single coverage

### Adding coverage heatmap

### Multiple coverage

### Coverage heatmap at TSS

## Expression heatmap

## Correlation heatmap

## Linking several heatmaps

# Interactive genome tracks



## Single locus

## Input Data
A trackview requires three components to display properly.

1. The view range: a GRanges object describing the region to be displayed
2. Annotation data: a list containing annotation features that will be rendered
3. Coverage data: a GRanges object that contains the coverage to be displayed

### View Range
Browserly needs to know the range to plot. This is used to pull the correct annotation information. The function `get_view_range` makes the appropriate object for use with browserly. If passed an OrganismDbi object, `get_view_range` can also be used to get the range by gene symbol. 

```{r view_range}
view_range <- get_view_range(chr = "chr21", start=9688000, end=9690000)
view_range
```

The range can also be extended if desired. This is a particularly useful function when creating uniformly sized views around disparate regions. For example, the TSS of many genes can be made as 1bp GRanges and extended to be of equal length.

The range can be extended symmetrically on both sides
```{r extend range symmetric}
extend_grange(view_range, 50000)
```

or asymmetrically by passing a vector
```{r extend range asymmetric}
extend_grange(view_range, c(1000, 5000))
```

### Annotation data
We pre-load the annotation information from the TxDB into an easily searched list. This saves time when repeated calls are made to find the genomic ranges and features that are used in plotting coverage information. 
The variable `tx_data` contains several lists that a indexed by transcript
feature (e.g. intron, utr5). It also keeps track of the seqlevelstyle used
in the database.

```{r Load Annotation reference}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
tx_data <- load_tx_data(txdb)
print(names(tx_data))
```

**Note:** We also use the annotation helper function `match_tx_to_gene` to zip together OrgDB and TxDB objects by entrez id. This allows genes to be searched by gene symbol, which is often more human friendly.
``` {r match transcripts to genes}
library(org.Hs.eg.db)
symbol_table <- match_tx_to_gene(txdb, "org.Hs.eg")
head(symbol_table)
```

### Coverage information.
Users are expected to provide their own data for plotting. Data should be provided as a GRanges object, where the metadata columns (mcols) contain the coverage values in each range.

The utility function `make_coverage_tracks` is provided for convenience, and has a few helpful features. It takes a list of coverage files (e.g. bigwig, bam) and the `view_range` to be displayed. Optionally, the user can provide corresponding names for the samples (e.g. treatment condition, antibody), and a scaling factor for the coverage data (e.g. library size).

```{r Get coverage}
sf <- rep(100,6)
# Get the coverage
cvg <- make_coverage_tracks(bamfiles[c(1,4,6)], 
                            target_range = view_range, 
                            scaling_factors = sf[c(1,4,6)])
head(cvg)
```

## Standard plot
A common representation of coverage is a trackview. Browserly makes it simple to 
create an interactive trackview that looks similar to other packages, such as Gviz.
```{r Demo_stacking}
plot_single_locus(view_range, tx_data, cvg, stacking='squish', type="scatter")
```

By default, browserly densely stacks all overlapping transcripts to save visual space. Future versions may allow for the display of "canonical" transcripts. All y axes default to being on the same scale, but this can also be disabled.
```{r Demo_line}
#plot_single_locus(view_range, tx_data, cvg, type = "scatter", sync_y = FALSE)
```

## Heatmap of coverage
Browserly automatcially determines the type of plot for showing coverage. When many
samples are displayed, browserly defaults to using a heatmap. This saves vertical space, without losing information.
```{r Demo_hm}
plot_single_locus(view_range, tx_data, cvg)
```
```{r}
tr_fn <- make_browserly_function(bamfiles[c("Ctcf","Znf143","Suz12")], tx_data = tx_data,
                                cvg_scaling = 100)
```


## Multiple locus

## Adding expression information

## Log scale???

# Apps linking heatmaps to genome tracks


```{r}

```





