---
title: "RNA Heatmap Example"
author: "Alicia Schep"
date: "7/12/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r simple_heatmap, message = FALSE}
library(plotly)
library(Biobase)
library(chipVis)
library(iHeatmapR)
options(ucscChromosomeNames=FALSE)

e <- readRDS("~/workspace/BAP1_Data/esetRPKM.RDS")

e <- e[which(substr(fData(e)$chr,1,3) == "chr"),]
fData(e)$chr <- sapply(fData(e)$chr, function(x) substr(x, 4, nchar(x)))
e <- e[which(fData(e)$chr %in% c(1:22,"X")),]

#source("~/workspace/chipVis/R/signal_heatmap.R")
#source("~/workspace/chipVis/R/plot_utils.R")

p1 = simple_heatmap(log10(exprs(e)[1:500,] + 1), 
                     y = as.character(fData(e)$symbol[1:500]), 
                     x = as.character(pData(e)$TREATMENT_NAME), 
                     row_order = "kmeans", 
                     col_order = "hclust", 
                     row_k = 8, 
                     col_k = 4,
                     y_pos = "none") %>%
  add_col_groups(pData(e)$TREATMENT_NAME, name = "Genotype", 
                  colorscale = dcolorscale(4,palette="Set2"), side = "bottom")

plot_iHeatmap(p1) %>% layout(margin = list(b = 200))
```


# Chip-seq

```{r}
BiocParallel::register(BiocParallel::MulticoreParam(4))

file.sample <- "~/workspace/BAP1_Data/BAP1_june_2016_samples.txt"
file.pair <- "~/workspace/BAP1_Data/BAP1_june_2016_pairs.txt"

samples <- read.table(file.sample,as.is=TRUE,header=TRUE)
samples_exp <- expandSampleTable(samples)

pairs <- read.table(file.pair,as.is=TRUE,header=TRUE)
pairs_exp <- expandPairedTable(pairs)

bam_files <- samples_exp$File.bam
names(bam_files) <- samples$Genotype

cvg_files <- samples_exp$File.bw
names(cvg_files) <- samples$Genotype

promoter_ranges <- with(fData(e), GenomicRanges::GRanges(chr,
                                                         IRanges::IRanges(start,start),
                                                         strand = str,
                                                         symbol = symbol)) %>%
  GenomicRanges::promoters(upstream = 2500, downstream = 2500)


promoter_counts <- simplify2array(BiocParallel::bplapply(bam_files,
                                                         bamsignals::bamCount,
                                                         promoter_ranges[1:500]))
 
p1 %>% 
  add_simple_heatmap(promoter_counts, col_order = "hclust", k_col = 4, scale_method = "normalize") %>%
  plot_iHeatmap()  %>% 
  layout(margin = list(b = 200))

```
