---
title: "browser_ly_testing"
author: "Justin Finkle"
date: "7/27/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(TxDb.Hsapiens.BioMart.igis)
library(plotly)
library(org.Hs.eg.db)
library(chipVis)
```


```{r}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")
org <- "human"
genome <- "GRCh38"


T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)
```

```{r}
tx_data <- load_tx_data(TxDb.Hsapiens.BioMart.igis)
symbol_table <- match_tx_to_gene(TxDb.Hsapiens.BioMart.igis, 'org.Hs.eg')
```

```{r}
# This happens to be the gene region of AGGF1 and ZBED3
view_range <- get_view_range(chr = 5, start=76320000, end=76400000)

# Get the coverage
input <- fi[2, 'Sample.Control']
input_idx <- which(fi$Sample.Name==input)
idx <- sapply(T47D[c(1,2,4,5)], function(x){x[[1]]})
idx <- c(1, idx)
names(idx)[[1]] <- "ER_input"
shortened_names <- names(idx)

# Get scaling factor
sf <- get_scaling_factor(fi[idx,])/1000000
sf <- sf[2,]

# Get the coverage
cvg <- make_coverage_tracks(fi[idx, "File.bw"], 
                            target_range = view_range, 
                            sample_names = shortened_names, 
                            scaling_factors = sf)
```


```{r Demo_hm}
plot_single_locus(view_range, tx_data, cvg)
```

```{r Demo_line}
plot_single_locus(view_range, tx_data, cvg, type = "scatter")
```

```{r Demo_stacking}
plot_single_locus(view_range, tx_data, cvg, stacking='squish', type="scatter")
```

# Plotting multiple genes for several conditions
```{r}
# set the gene list
genes <- c('RNF2', 'E2F1', 'GREB1', 'BAP1')

# Select the indices of the chip-seq experiments
# c_idx stands for conditions index

# Set how far upstream and downstream  you want to view from the TSS
extension <- 50000
conditions <- names(idx)
cvg_files <- fi[idx, "File.bw"]

centered_genes <- lapply(genes, function(gene){
  gene_data <- get_centered_gene_info(gene = gene,
                                      extension = extension,
                                      cvg_files = cvg_files,
                                      conditions = conditions,
                                      scaling_factor = sf,
                                      tx_data = tx_data,
                                      symbol_table = symbol_table)
  return(gene_data)
})
cvg_list <- GRangesList(lapply(centered_genes, function(x) {x$gene_cvg}))
names(cvg_list) <- genes
tx_list <- GRangesList(lapply(centered_genes, function(x) {x$gene_tx}))
names(tx_list) <- genes

```

```{r}
plot_multiple_genes(genes = genes,
                    centered_cvg = cvg_list,
                    centered_tx = tx_list)
```