---
title: "browser_ly_testing"
author: "Justin Finkle"
date: "7/27/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(TxDb.Hsapiens.BioMart.igis)
library(plotly)
library(org.Hs.eg.db)
library(chipVis)
```


```{r}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")
org <- "human"
genome <- "GRCh38"


T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)
```

```{r}
tx_data <- load_tx_data(TxDb.Hsapiens.BioMart.igis)
symbol_table <- match_tx_to_gene(TxDb.Hsapiens.BioMart.igis, 'org.Hs.eg')
```

```{r}
# This happens to be the gene region of AGGF1 and ZBED3
view_range <- get_view_range(chr = 5, start=76320000, end=76400000)

# Get the coverage
input <- fi[2, 'Sample.Control']
input_idx <- which(fi$Sample.Name==input)
idx <- sapply(T47D[c(1,2,4,5)], function(x){x[[1]]})
idx <- c(1, idx)
names(idx)[[1]] <- "ER_input"
shortened_names <- names(idx)

# Get scaling factor
sf <- get_scaling_factor(fi[idx,])/1000000
sf <- sf[2,]

# Get the coverage
cvg <- make_coverage_tracks(fi[idx, "File.bw"], 
                            target_range = view_range, 
                            sample_names = shortened_names, 
                            scaling_factors = sf)
```


```{r Demo_hm}
plot_single_locus(view_range, tx_data, cvg)
```

```{r Demo_line}
plot_single_locus(view_range, tx_data, cvg, type = "scatter")
```

```{r Demo_stacking}
plot_single_locus(view_range, tx_data, cvg, stacking='squish', type="scatter")
```

# Plotting multiple genes for several conditions
```{r}
# set the gene list
genes <- c('RNF2', 'E2F1', 'GREB1', 'BAP1')

# Select the indices of the chip-seq experiments
# c_idx stands for conditions index
c_idx <- c(2, 6, 10, 14, 19)

# Set how far upstream and downstream  you want to view from the TSS
extension <- 50000
```

```{r Multiple Genes, echo=FALSE}
lib_size <- get_scaling_factor(fi[c_idx,])[2,]/1000000
conditions <- fi$Genotype[c_idx]
p_list <- list()
a_list <- list()
r_list <- list()
score_max <- 0
for(gene in genes){
  sl <- ifelse(gene == genes[1], TRUE, FALSE)
  p <- plot_ly(type = 'scatter', showlegend=sl, name=gene)
  x <- which(symbol_table$symbol == gene)
  
  # Get the range of the gene
  cur_tss <- ifelse(as.character(symbol_table$strand[x]) == "-",
                    symbol_table$end[x],
                    symbol_table$start[x])
  cur_range <- get_view_range(chr = symbol_table$chr[x],
                              start = cur_tss - extension,
                              end = cur_tss + extension,
                              strand = as.character(symbol_table$strand[x]))
  # Get the coverage in the range
  cur_cvg <- make_coverage_tracks(fi[c_idx, "File.bw"], 
                            target_range = cur_range, 
                            sample_names = conditions, 
                            scaling_factors = lib_size)
  # Get max score for scaling y_axes
  score_max <- max(max(as.data.frame(mcols(cur_cvg)), score_max))
  cvg_df <- biovizBase::mold(cur_cvg)
  
  # Get the relative distances
  cvg_df$midpoint <- cvg_df$midpoint - cur_tss
  
  # Add the coverage as traces
  for(cond in conditions){
    p <- p %>% add_trace(x = cvg_df$midpoint, y = cvg_df[[cond]], mode='lines',
                         name = cond, color=cond)
  }
  
  #Save the annotation information
  cur_tx <- get_tx_annotation(range=cur_range, tx_data = tx_data)
  cur_tx$stepping <- -1
  cur_tx <- biovizBase::mold(cur_tx)
  cur_tx$start <- cur_tx$start - cur_tss
  cur_tx$end <- cur_tx$end - cur_tss
  cur_tx$midpoint <- cur_tx$midpoint - cur_tss
  a_list[[gene]] <- cur_tx
  start(cur_range) <- start(cur_range) - cur_tss
  end(cur_range) <- end(cur_range) - cur_tss
  r_list[[gene]] <- cur_range
  
  # Add invisible annotation handles
  h_features <- unique(cur_tx$feature)
  for(h_feature in h_features){
    h <- cur_tx[cur_tx$feature == h_feature, ]
    p <- p %>% add_trace(x=h$midpoint, 
                         y=h$stepping, 
                         opacity=0,
                         text = h$transcript, 
                         name = h_feature,
                         # color = h_feature,
                         # colors = rep('grey', length.out = length(h_features)),
                         hoverinfo="x+name+text",
                         marker = list(color='grey'),
                         showlegend=FALSE)
  }
  p_list[[gene]] <- p
}
sp <- subplot(p_list, nrows=length(p_list), shareX = TRUE)

```

```{r Add annotations, echo=FALSE}
trace_names <- sapply(sp$x$data, function(x){x$name})
trace_axes <- lapply(sp$x$data, function(x) {gsub("y", "yaxis", x$yaxis)})
names(trace_axes) <- trace_names  
sp2 <- sp
for(gene in genes){
  yax <- trace_axes[[gene]]
  sp2 <- add_tx_shapes(sp2, a_list[[gene]], r_list[[gene]], yax)
  sp2$x$layout[[yax]][['title']] <- gene
  sp2$x$layout[[yax]][['range']] <- c(-1.5, score_max)
}
sp2$x$layout$showlegend <- TRUE
sp2$x$layout$xaxis$title <- 'Distance to TSS'
sp2
```