---
title: "Untitled"
runtime: shiny
output: html_document
---

```{r setup, include=FALSE, echo=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, autodep=TRUE)
library(gChipseq)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(plotly)
library(iHeatmapR)
BiocParallel::register(BiocParallel::MulticoreParam(4, progressbar = FALSE))
```

``` {r Load Data}

fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000

rpkmEset <- readRDS("/gne/research/workspace/schepa/Carroll_Data/GSE68358.rpkmEset.with_fdata.Rds")

tss_ranges <- with(fData(rpkmEset), GenomicRanges::GRanges(chr,
                                               IRanges::IRanges(start = ifelse(strand == "-",
                                                                               end,
                                                                               start),
                                                                width = 1),
                                               strand = strand,
                                               symbol = symbol,
                                               entrez = entrez))


```
  
 
```{r}
T47D_eset <- rpkmEset[,grep("T47D",pData(rpkmEset)$celltype)]

T47D_exprs = log10(exprs(T47D_eset)+1)
T47D_var = apply(T47D_exprs,1, sd, na.rm= TRUE)
high_var = which(T47D_var > quantile(T47D_var,0.95,na.rm= TRUE))
treat_groups = stringr::str_trim(sapply(stringr::str_split(pData(T47D_eset)$treatment,":"), 
                      function(x) x[[2]]))

rna_hm <- rna_heatmap(T47D_exprs[high_var,], col_groups = treat_groups,
            x = pData(T47D_eset)$title, y = tss_ranges$symbol[high_var],
            y_title = NULL, x_title = NULL) 

# rna_hm %>% plot_iHeatmap() %>% layout(margin = list(b = 250)) %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/rna_hm.pdf")

```

  
#Get T47D Progesterone  
  
```{r MCF7_tss_cvg}
T47D = list(
  T47D_ER_Full_Media_3hr  = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  T47D_ER_Progesterone_3hr = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  T47D_ER_R5020_3hr = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  T47D_PR_Full_Media_3hr  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  T47D_PR_Progesterone_3hr = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  T47D_PR_R5020_3hr = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  T47D_p300_Full_Media_3hr  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  T47D_p300_Progesterone_3hr = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  T47D_p300_R5020_3hr = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)


# T47D = list(
#   T47D_ER_3hr_Vehicle  = which(fi$source_name == "T47D_ER_3hr_Vehicle"),
#   T47D_ER_3hr_E2 = which(fi$source_name == "T47D_ER_3hr_E2"),
#   T47D_ER_3hr_PG = which(fi$source_name == "T47D_ER_3hr_PG"),
#   T47D_ER_3hr_E2_PG = which(fi$source_name == "T47D_ER_3hr_E2+PG"),
#   
#   T47D_PR_3hr_Vehicle  = which(fi$source_name == "T47D_PR_3hr_Vehicle"),
#   T47D_PR_3hr_E2 = which(fi$source_name == "T47D_PR_3hr_E2"),
#   T47D_PR_3hr_PG = which(fi$source_name == "T47D_PR_3hr_PG"),
#   T47D_PR_3hr_E2_PG = which(fi$source_name == "T47D_PR_3hr_E2+PG")
#   
# )

tss_cvg_mats <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = tss_ranges[high_var],
                                  binsize = 100,
                                  up = 5000,
                                  down = 5000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})

tr <- transcriptsBy(TxDb.Hsapiens.BioMart.igis, by = "gene") %>% unlist()
tr$entrez <- stringr::str_extract(names(tr),"(?<=:)(.*)")
tr$symbol <- annotate::getSYMBOL(tr$entrez, data='org.Hs.eg')
names(tr) = NULL
tr <- resize(tr, fix = "start", width = 1)

tss_cvg_mats2 <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = tr[which(tr$entrez %in% tss_ranges$entrez[high_var])],
                                  binsize = 25,
                                  up = 1000,
                                  down = 1000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})

# 
# tss <- gChipseq::tssFromTxdb(txdb = TxDb.Hsapiens.BioMart.igis)
# #unique_tss <- unique(tss) #Remove same positions
# 
# peaks <- lapply( fi$File.macs[unlist(T47D)], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
# peaks <- peaks[!sapply(peaks, is.null)]
# 
# merged_peaks <- reduce(do.call(c, peaks))
# 
# tss_in_peaks <- subsetByOverlaps(tss_ranges, merged_peaks, maxgap = 0)
# 
# tss_cvg_mats <- lapply(T47D, function(x){
#   tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
#                                   ranges = tss_ranges[high_var],
#                                   binsize = 25,
#                                   up = 1000,
#                                   down = 1000)
#   out <- Reduce("+", tmp) / length(tmp)
#   return(out)
# })

```



```{r}
new_single_coverage_heatmap(tss_cvg_mats[[1]], row_order = "signal", y = tss_ranges$symbol[high_var])
```

```{r}
T47D_ER_tss_cvg_mats = tss_cvg_mats[1:3]
names(T47D_ER_tss_cvg_mats) = c("Control","Pr","Syn. Pr")
multi_coverage_heatmap(T47D_ER_tss_cvg_mats, row_order = "kmeans", y = tss_in_peaks$symbol, k = 3)
```

```{r}
MCF7_PR_tss_cvg_mats = tss_cvg_mats[4:6]
names(MCF7_PR_tss_cvg_mats) = c("Control","Pr","Syn. Pr")
multi_coverage_heatmap(MCF7_PR_tss_cvg_mats, row_order = "kmeans", y = tss_in_peaks$symbol, k = 3)
```


```{r}
peaks <- lapply( fi$File.macs[unlist(T47D)], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
peaks <- peaks[!sapply(peaks, is.null)]

merged_peaks <- lapply(T47D,
                       function(x){
                         peaks <- lapply( fi$File.macs[x], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
                         peaks <- peaks[!sapply(peaks, is.null)]
                         reduce(do.call(c, peaks), min.gapwidth = 500)
                         })
                       

er_progesterone_peaks_cvg_mats <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = merged_peaks[[2]][1:5000],
                                  binsize = 50,
                                  up = 1000,
                                  down = 1000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})
#merged_peaks <- reduce(do.call(c, peaks))

T47D_scale_factors = sapply(T47D, function(x) mean(sf[x]))

```

```{r}
ag1 = aggregate_profile_plot(er_progesterone_peaks_cvg_mats[1:3],xlab = "",
                            ylab = "ER ChIP-Signal")
ag2 = aggregate_profile_plot(er_progesterone_peaks_cvg_mats[4:6],xlab = "Position relative to ER Peaks<br> in T47D + Progesterone",
                            ylab = "PR ChIP-Signal")

ag1 %>% export(file = "~/workspace//Carroll_Data/Figures/aggregate_profiles1.pdf")
# ag2 %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/aggregate_profiles2.pdf")


```

```{r}
mchm <- multi_coverage_heatmap(er_progesterone_peaks_cvg_mats[1:3], row_order = "kmeans", k = 3, cluster_by = "first")

# mchm %>% plot_iHeatmap %>% layout(margin = list(b = 250)) %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/multi_coverage.pdf")

```


