---
title: "Untitled"
runtime: shiny
output: html_document
---

```{r setup, include=FALSE, echo=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, autodep=TRUE)
library(gChipseq)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(plotly)
library(iHeatmapR)
BiocParallel::register(BiocParallel::MulticoreParam(4, progressbar = FALSE))
```

``` {r Load Data}
info_dir <- "/gne/research/workspace/rinaldj1/projects/bulkChipPipeline/data/CARROLL_PR/data/"
s_file <- paste(info_dir, "samples.txt", sep='')
p_file <- paste(info_dir, "pairs.txt", sep='')
fi <- chipVis::get_chip_file_info(s_file, p_file)
org <- "human"
genome <- "GRCh38"

anno <- readr::read_csv("/gne/research/workspace/rinaldj1/projects/bulkChipPipeline/data/CARROLL_PR/datDex.csv")[,c("run","sample_attribute")]

sample_attributes = lapply(strsplit(anno$sample_attribute, split = "||", fixed = TRUE),
                           function(x) setNames(lapply(x, function(y) stringr::str_trim(strsplit(y,":")[[1]][2])),
                                                stringr::str_trim(sapply(x, function(y) strsplit(y,":")[[1]][1]))))

sample_attributes_df <- plyr::ldply(sample_attributes, data.frame, stringsAsFactors = FALSE)

fi_attributes <- as.data.frame(t(sapply(fi$Sample.Name, function(x){
  if (x %in% anno$run){
    return(sample_attributes_df[which(anno$run == x), ])
  } else{
    return(sample_attributes_df[grep(x,sample_attributes_df$source_name)[1], ])
  }
})))

fi <- cbind(fi, fi_attributes)

# Get scaling factors based on library size
sf <- chipVis::get_scaling_factor(fi)[2,]/1000000

```
  
  
```{r}
library(Homo.sapiens)
rpkmEset <- readRDS("/gne/research/workspace/schepa/Carroll_Data/GSE68358.rpkmEset.Rds")

fd <- AnnotationDbi::select(Homo.sapiens, 
             rownames(exprs(rpkmEset)), 
             c("TXSTART","TXEND","TXCHROM","TXSTRAND","SYMBOL","ENSEMBL"),
             keytype="ENTREZID")

library(dplyr)
fd2 <- fd %>% group_by(ENTREZID) %>% summarize(start = min(TXSTART),
                                                 end = max(TXEND),
                                                 chr = first(TXCHROM),
                                                 strand = first(TXSTRAND),
                                                 symbol = first(SYMBOL),
                                                 ensembl = first(ENSEMBL))

fd2 <- fd2[which(substr(fd2$chr,1,3) == "chr"),]
fd2$chr <- sapply(fd2$chr, function(x) substr(x, 4, nchar(x)))
fd2 <- fd2[which(fd2$chr %in% c(1:22,"X")),]

rpkmEset <- rpkmEset[fd2$ENTREZID,]

tss_ranges <- with(fd2, GenomicRanges::GRanges(chr,
                                               IRanges::IRanges(start = ifelse(strand == "-",
                                                                              end,
                                                                              start),
                                                               width = 1),
                                               strand = strand,
                                               symbol = symbol,
                                               entrez = ENTREZID,
                                               ensembl = ensembl))
T47D_eset <- rpkmEset[,grep("T47D",pData(rpkmEset)$celltype)]

```
  
  
```{r}
T47D_exprs = log10(exprs(T47D_eset)+1)
T47D_var = apply(T47D_exprs,1, sd, na.rm= TRUE)
high_var = which(T47D_var > quantile(T47D_var,0.95,na.rm= TRUE))
treat_groups = stringr::str_trim(sapply(stringr::str_split(pData(T47D_eset)$treatment,":"), 
                      function(x) x[[2]]))

rna_hm <- rna_heatmap(T47D_exprs[high_var,], col_groups = treat_groups,
            x = pData(T47D_eset)$title, y = tss_ranges$symbol[high_var],
            y_title = NULL, x_title = NULL) 

# rna_hm %>% plot_iHeatmap() %>% layout(margin = list(b = 250)) %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/rna_hm.pdf")

```

  
#Get T47D Progesterone  
  
```{r MCF7_tss_cvg}
T47D = list(
  T47D_ER_Full_Media_3hr  = which(fi_attributes$source_name == "T47D_ER_Full_Media_3hr"),
  T47D_ER_Progesterone_3hr = which(fi_attributes$source_name == "T47D_ER_Progesterone_3hr"),
  T47D_ER_R5020_3hr = which(fi_attributes$source_name == "T47D_ER_R5020_3hr"),
  
  T47D_PR_Full_Media_3hr  = which(fi_attributes$source_name == "T47D_PR_Full_Media_3hr"),
  T47D_PR_Progesterone_3hr = which(fi_attributes$source_name == "T47D_PR_Progesterone_3hr"),
  T47D_PR_R5020_3hr = which(fi_attributes$source_name == "T47D_PR_R5020_3hr"),
  
  T47D_p300_Full_Media_3hr  = which(fi_attributes$source_name == "T47D_p300_Full_Media_3hr"),
  T47D_p300_Progesterone_3hr = which(fi_attributes$source_name == "T47D_p300_Progesterone_3hr"),
  T47D_p300_R5020_3hr = which(fi_attributes$source_name == "T47D_p300_R5020_3hr")
  
)


# T47D = list(
#   T47D_ER_3hr_Vehicle  = which(fi_attributes$source_name == "T47D_ER_3hr_Vehicle"),
#   T47D_ER_3hr_E2 = which(fi_attributes$source_name == "T47D_ER_3hr_E2"),
#   T47D_ER_3hr_PG = which(fi_attributes$source_name == "T47D_ER_3hr_PG"),
#   T47D_ER_3hr_E2_PG = which(fi_attributes$source_name == "T47D_ER_3hr_E2+PG"),
#   
#   T47D_PR_3hr_Vehicle  = which(fi_attributes$source_name == "T47D_PR_3hr_Vehicle"),
#   T47D_PR_3hr_E2 = which(fi_attributes$source_name == "T47D_PR_3hr_E2"),
#   T47D_PR_3hr_PG = which(fi_attributes$source_name == "T47D_PR_3hr_PG"),
#   T47D_PR_3hr_E2_PG = which(fi_attributes$source_name == "T47D_PR_3hr_E2+PG")
#   
# )

tss <- gChipseq::tssFromTxdb(txdb = TxDb.Hsapiens.BioMart.igis)
#unique_tss <- unique(tss) #Remove same positions

peaks <- lapply( fi$File.macs[unlist(T47D)], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
peaks <- peaks[!sapply(peaks, is.null)]

merged_peaks <- reduce(do.call(c, peaks))

tss_in_peaks <- subsetByOverlaps(tss_ranges, merged_peaks, maxgap = 0)

tss_cvg_mats <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = tss_ranges[high_var],
                                  binsize = 25,
                                  up = 1000,
                                  down = 1000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})

```



```{r}
new_single_coverage_heatmap(tss_cvg_mats[[1]], row_order = "signal", y = tss_ranges$symbol)
```

```{r}
T47D_ER_tss_cvg_mats = tss_cvg_mats[1:3]
names(T47D_ER_tss_cvg_mats) = c("Control","Pr","Syn. Pr")
multi_coverage_heatmap(T47D_ER_tss_cvg_mats, row_order = "kmeans", y = tss_in_peaks$symbol, k = 3)
```

```{r}
MCF7_PR_tss_cvg_mats = tss_cvg_mats[4:6]
names(MCF7_PR_tss_cvg_mats) = c("Control","Pr","Syn. Pr")
multi_coverage_heatmap(MCF7_PR_tss_cvg_mats, row_order = "kmeans", y = tss_in_peaks$symbol, k = 3)
```


```{r}
peaks <- lapply( fi$File.macs[unlist(T47D)], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
peaks <- peaks[!sapply(peaks, is.null)]

merged_peaks <- lapply(T47D,
                       function(x){
                         peaks <- lapply( fi$File.macs[x], readMacsPeaks, fc=5, minTags=10, fdr=0.01)
                         peaks <- peaks[!sapply(peaks, is.null)]
                         reduce(do.call(c, peaks), min.gapwidth = 500)
                         })
                       

er_progesterone_peaks_cvg_mats <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = merged_peaks[[2]][1:5000],
                                  binsize = 50,
                                  up = 1000,
                                  down = 1000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})
#merged_peaks <- reduce(do.call(c, peaks))

T47D_scale_factors = sapply(T47D, function(x) mean(sf[x]))

```

```{r}
ag1 = aggregate_profile_plot(er_progesterone_peaks_cvg_mats[1:3],xlab = "",
                            ylab = "ER ChIP-Signal")
ag2 = aggregate_profile_plot(er_progesterone_peaks_cvg_mats[4:6],xlab = "Position relative to ER Peaks<br> in T47D + Progesterone",
                            ylab = "PR ChIP-Signal")

ag1 %>% export(file = "~/workspace//Carroll_Data/Figures/aggregate_profiles1.pdf")
# ag2 %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/aggregate_profiles2.pdf")


```

```{r}
mchm <- multi_coverage_heatmap(er_progesterone_peaks_cvg_mats[1:3], row_order = "kmeans", k = 3, cluster_by = "first")

# mchm %>% plot_iHeatmap %>% layout(margin = list(b = 250)) %>% export(file = "/gne/research/workspace/schepa/Carroll_Data/Figures/multi_coverage.pdf")

```


