---
title: "Feature Extraction"
author: "Justin Finkle"
date: "8/17/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(org.Hs.eg.db)
```

```{r Load Data}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000
```

```{r Match transcripts with symbol, echo = FALSE}
tr <- transcriptsBy(TxDb.Hsapiens.BioMart.igis, by = "gene") %>% unlist()
tr$entrez <- stringr::str_extract(names(tr),"(?<=:)(.*)")
tr$symbol <- annotate::getSYMBOL(tr$entrez, data='org.Hs.eg')
names(tr) = NULL
trdf <- as.data.frame(tr)

truniq <- trdf %>% group_by(entrez) %>% summarize(chr = first(seqnames),
                                                  start = min(start),
                                                  end = min(end),
                                                  strand = first(strand),
                                                  symbol = first(symbol))
```

```{r}
symbol <- "GREB1"
sym_idx <- which(truniq$symbol == symbol)
symbol_range <- get_view_range(chr = truniq$chr[sym_idx], start = truniq$start[sym_idx],
                               end = truniq$end[sym_idx])
symbol_range <- extend_grange(symbol_range, 500000)

```

```{r}
n_recurse_tiles <- 3
width(recursive_tile_range(symbol_range, n = n_recurse_tiles))
scaled_range <- log_scale_range(gr = symbol_range, n = n_recurse_tiles)
```


