---
title: "Feature Extraction"
author: "Justin Finkle"
date: "8/17/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache = TRUE)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(org.Hs.eg.db)
library(dplyr)
library(plotly)
library(iHeatmapR)
```

```{r Load Data}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000

T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)
```

```{r Match transcripts with symbol, echo = FALSE}
tr <- transcriptsBy(TxDb.Hsapiens.BioMart.igis, by = "gene") %>% unlist()
tr$entrez <- stringr::str_extract(names(tr),"(?<=:)(.*)")
tr$symbol <- annotate::getSYMBOL(tr$entrez, data='org.Hs.eg')
names(tr) = NULL
trdf <- as.data.frame(tr)

truniq <- trdf %>% group_by(entrez) %>% dplyr::summarize(chr = first(seqnames),
                                                  start = min(start),
                                                  end = max(end),
                                                  strand = first(strand),
                                                  symbol = first(symbol))
```

```{r}
symbol <- "NRIP1"
sym_idx <- which(truniq$symbol == symbol)
symbol_range <- get_view_range(chr = truniq$chr[sym_idx], start = truniq$start[sym_idx],
                               end = truniq$end[sym_idx], strand = as.character(truniq$strand[sym_idx]))
```

```{r}
n_recurse_tiles <- 10
pseudo_tss <- ifelse(strand(symbol_range)=="-", end(symbol_range), start(symbol_range))
scaled_range <- log_scale_range(center = pseudo_tss, 
                                chromosome = seqnames(symbol_range),
                                n = n_recurse_tiles, 
                                min_bin = 100)
```

# Log-Scale for NRIP1
The log scale gives us a good dynamic range for seeing peaks. The plot shows the 
average peak height per bin. Peaks don't really show up until 100kb from the TSS
```{r}
log_peaks <- peaks_per_sample(T47D, fi, scaled_range)
all_peaks <- scaled_range
mcols(all_peaks) <- Reduce(cbind, log_peaks)
df <- biovizBase::mold(all_peaks)
hm_vals <- apply(t(as.matrix(as.data.frame(mcols(all_peaks)))), 2, rev)+1
# hm_vals <- matrix(rnorm(length(hm_vals)), nrow=nrow(hm_vals))
p <- plot_ly(z=hm_vals, y=rownames(hm_vals),type='heatmap', colorscale = continuous_colorscale("Purples")(hm_vals),
               hoverinfo='x+z')
lines <- make_lines(symbol_range = symbol_range, scaled_range = scaled_range, hm_vals = hm_vals, base = n_recurse_tiles)
p %>% layout(shapes=lines$lines,
             xaxis=list(tickvals = lines$vals, ticktext = names(lines$vals), tickmode='array',
                        zeroline=FALSE, showline=TRUE, mirror=TRUE))
```

# Linear Scale
The plot below shows the same range for peaks, but on a linear scale. Bins are 
around 27kb long. This makes it look like peaks are right next to the TSS, which
is not the case
```{r Large Scale}
tiled_range <- tile(reduce(scaled_range), n = length(scaled_range))[[1]]
even_peaks <- peaks_per_sample(T47D, fi, tiled_range)
all_peaks <- tiled_range
mcols(all_peaks) <- Reduce(cbind, even_peaks)
df <- biovizBase::mold(all_peaks)
hm_vals <- apply(t(as.matrix(as.data.frame(mcols(all_peaks)))), 2, rev)+1
# hm_vals <- matrix(rnorm(length(hm_vals)), nrow=nrow(hm_vals))
p <- plot_ly(z=hm_vals, y=rownames(hm_vals),type='heatmap', colorscale = continuous_colorscale("Purples")(hm_vals),
               hoverinfo='x+z')
lines <- make_lines(symbol_range = symbol_range, scaled_range = tiled_range, hm_vals = hm_vals, base = n_recurse_tiles)
p %>% layout(xaxis=list(tickvals = lines$vals, ticktext = names(lines$vals), tickmode='array',
                        zeroline=FALSE, showline=TRUE, mirror=TRUE))
```

# Linear Scale, small bins
Here we also use a linear scale, but with smaller bins of 1kb. We no longer have
the issue of thinking peaks are close to the TSS, however visually it is hard to
see much signal. If you zoom to +100kb, you'll see that there are indeed strong
peaks. Overall though, the heatmap is very sparse, and it it isn't clear where
the user should zoom to see the strong peaks. 
```{r 1000bp bins}
tiled_range <- tile(reduce(scaled_range), width = 1000)[[1]]
even_peaks <- peaks_per_sample(T47D, fi, tiled_range)
all_peaks <- tiled_range
mcols(all_peaks) <- Reduce(cbind, even_peaks)
df <- biovizBase::mold(all_peaks)
hm_vals <- apply(t(as.matrix(as.data.frame(mcols(all_peaks)))), 2, rev)+1
# hm_vals <- matrix(rnorm(length(hm_vals)), nrow=nrow(hm_vals))
p <- plot_ly(z=hm_vals, y=rownames(hm_vals),type='heatmap', colorscale = continuous_colorscale("Purples")(hm_vals),
               hoverinfo='x+z')
lines <- make_lines(symbol_range = symbol_range, scaled_range = tiled_range, hm_vals = hm_vals, base = n_recurse_tiles)
p %>% layout(xaxis=list(tickvals = lines$vals, ticktext = names(lines$vals), tickmode='array',
                        zeroline=FALSE, showline=TRUE, mirror=TRUE))
```