---
title: "Feature Extraction"
author: "Justin Finkle"
date: "8/17/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(org.Hs.eg.db)
library(dplyr)
```

```{r Load Data}
fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000

T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)
```

```{r Match transcripts with symbol, echo = FALSE}
tr <- transcriptsBy(TxDb.Hsapiens.BioMart.igis, by = "gene") %>% unlist()
tr$entrez <- stringr::str_extract(names(tr),"(?<=:)(.*)")
tr$symbol <- annotate::getSYMBOL(tr$entrez, data='org.Hs.eg')
names(tr) = NULL
trdf <- as.data.frame(tr)

truniq <- trdf %>% group_by(entrez) %>% dplyr::summarize(chr = first(seqnames),
                                                  start = min(start),
                                                  end = max(end),
                                                  strand = first(strand),
                                                  symbol = first(symbol))
```

```{r}
symbol <- "GREB1"
sym_idx <- which(truniq$symbol == symbol)
symbol_range <- get_view_range(chr = truniq$chr[sym_idx], start = truniq$start[sym_idx],
                               end = truniq$end[sym_idx], strand = as.character(truniq$strand[sym_idx]))
```

```{r}
n_recurse_tiles <- 10
pseudo_tss <- ifelse(strand(symbol_range)=="-", end(symbol_range), start(symbol_range))
scaled_range <- log_scale_range(center = pseudo_tss, 
                                chromosome = seqnames(symbol_range),
                                n = n_recurse_tiles, 
                                min_bin = 100)
```

```{r}
range_peaks <- mclapply(seq_along(T47D), function(i, fi, scaled_range){
  x <- T47D[[i]]
  new_names <- paste0(rep(names(T47D)[[i]], length(x)), seq_along(x))
  tmp <- scaled_range
  for(n in seq_along(x)){
    peaks <- readMacsPeaks(fi$File.macs[x[[n]]])
    peaks_in_range <- avg_peak_in_range(gr = scaled_range, peaks_gr = peaks)
    mcols(tmp)[[new_names[n]]] <- mcols(peaks_in_range)[['avg_peak']]
  }
  return(mcols(tmp))
}, fi, scaled_range, mc.cores=8)
all_peaks <- scaled_range
mcols(all_peaks) <- Reduce(cbind, range_peaks)
```

```{r}
df <- biovizBase::mold(all_peaks)
hm_vals <- apply(t(as.matrix(as.data.frame(mcols(all_peaks)))), 2, rev)+1
hm_vals <- matrix(rnorm(length(hm_vals)), nrow=nrow(hm_vals))
p <- plot_ly(z=hm_vals, y=rownames(hm_vals),type='heatmap', colorscale = continuous_colorscale("Purples")(hm_vals),
               hoverinfo='x+z')
make_lines <- function(symbol_range, scaled_range, hm_vals, base){
  n_bins <- ncol(hm_vals)
  center <- (start(symbol_range) + end(symbol_range))/2
  left_width <- floor(log(cumsum(sort(width(scaled_range[1:floor(n_bins/2)]))), base))
  l_log_scale <- unique(left_width)[unique(left_width) >2 ]
  left_lines <- sapply(l_log_scale, function(x, widths){
    first(which(widths==x))
  }, widths=left_width)
  right_width <- floor(log(cumsum(sort(width(scaled_range[(floor(n_bins/2)+1):n_bins]))), base))
  r_log_scale <- unique(left_width)[unique(left_width) >2 ]
  right_lines <- sapply(r_log_scale, function(x, widths){
    first(which(widths==x))
  }, widths=right_width)
  center_line_x <- ncol(hm_vals)/2-0.5
  binsize <- floor(log(width(scaled_range), base))
  log_lines <- sort(c(-left_lines, 0, right_lines)) + center_line_x
  names(log_lines) <-  sapply(sort(c(-l_log_scale, 0, r_log_scale)), function(x) {
    if(x > 0){
      sign_val <- "+"
    } else if( x< 0){
      sign_val <- "-"
    } else {
      sign_val <- ""
    }
    return(paste0(sign_val, base, "<sup>", x, "</sup>")
  })
  line_list <- list()
  label_list <- list()
  for(ll in seq_along(log_lines)){
    x <- log_lines[[ll]]
    if(x == center_line_x){
      color <- 'red'
      label_list[[ll]] <- list(x = x, y=1, xref='x', yref='paper', text="center", showarrow=FALSE, yanchor='bottom')
    } else {
      color <- 'black'
      label_list[[ll]] <- list(showarrow=FALSE, text="")
    }
    line_list[[ll]] <- list(type='line', line=list(dash=3, color=color), x0=x, x1=x, y0=0, y1=1, yref='paper')
  }
  scale_list <- list(lines = line_list, labels = label_list, vals = log_lines)
  return(scale_list)
}
lines <- make_lines(symbol_range = symbol_range, scaled_range = scaled_range, hm_vals = hm_vals, base = n_recurse_tiles)
p %>% layout(shapes=lines$lines, annotations=lines$labels,
             xaxis=list(zeroline=TRUE, mirror=TRUE, linewidth=3,
                        tickvals = lines$vals, ticktext = names(lines$vals), tickmode='array'))
```

