---
title: "Interactive Poster"
author: "Alicia Schep and Justin Finkle"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
runtime: shiny
---

```{r setup, include=FALSE, echo=TRUE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, autodep=TRUE)
#library(gChipseq)
library(chipVis)
library(TxDb.Hsapiens.BioMart.igis3.0)
library(org.Hs.eg.db)
library(plotly)
library(dplyr)
library(iHeatmapR)
BiocParallel::register(BiocParallel::MulticoreParam(4, progressbar = FALSE))
```

``` {r Load Data, echo = FALSE}

fi <- readRDS("/gne/research/workspace/schepa/Carroll_Data/chip_file_info.Rds")

org <- "human"
genome <- "GRCh38"

# Get scaling factors based on library size
sf <- get_scaling_factor(fi)[2,]/1000000

rpkmEset <- readRDS("/gne/research/workspace/schepa/Carroll_Data/GSE68358.rpkmEset.with_fdata.Rds")

tss_ranges <- with(fData(rpkmEset), GenomicRanges::GRanges(chr,
                                               IRanges::IRanges(start = ifelse(strand == "-",
                                                                               end,
                                                                               start),
                                                                width = 1),
                                               strand = strand,
                                               symbol = symbol,
                                               entrez = entrez))

```
  
 
```{r, echo = FALSE}
T47D_eset <- rpkmEset[,grep("T47D",pData(rpkmEset)$celltype)]
treat_groups = stringr::str_trim(sapply(stringr::str_split(pData(T47D_eset)$treatment,":"), 
                      function(x) x[[2]])) %>% gsub("+","_",., fixed= TRUE) %>% 
  gsub("_3hr","",.)

library(limma)
library(edgeR)

dge <- DGEList(counts = exprs(T47D_eset)) %>% calcNormFactors()
design <- model.matrix(~ 0 +treat_groups)
colnames(design) = gsub("treat_groups", "", colnames(design))
v <- voom(dge, design=design)
fit <- lmFit(v, design)
contrasts <- makeContrasts(E2_Progesterone-E2, 
                           E2_R5020-E2, 
                           E2_R5020-E2_Progesterone, levels = design)
fit2 <- contrasts.fit(fit, contrasts) %>% eBayes()
results_Progesterone <- topTable(fit2, coef = 1, adjust = "BH")
results_R5020 <- topTable(fit2, coef = 2, adjust = "BH")


results <- decideTests(fit2, p.value = 0.01, lfc = 1)
sig <- which(apply(results,1, function(x) sum(x!=0) > 0))

T47D = list(
  ER_ab = which(fi$source_name == "T47D_ER_Full_Media_3hr"),
  ER_ab_Pr_treat = which(fi$source_name == "T47D_ER_Progesterone_3hr"),
  ER_ab_R5020_treat = which(fi$source_name == "T47D_ER_R5020_3hr"),
  
  PR_ab  = which(fi$source_name == "T47D_PR_Full_Media_3hr"),
  PR_ab_Pr_treat = which(fi$source_name == "T47D_PR_Progesterone_3hr"),
  PR_ab_R5020_treat = which(fi$source_name == "T47D_PR_R5020_3hr"),
  
  p300_ab  = which(fi$source_name == "T47D_p300_Full_Media_3hr"),
  p300_ab_Pr_treat  = which(fi$source_name == "T47D_p300_Progesterone_3hr"),
  p300_ab_R5020_treat = which(fi$source_name == "T47D_p300_R5020_3hr")
  
)

```

```{r}
col_group_names <- gsub(pattern = "Progesterone", replacement = "P4", treat_groups)
rna_hm <- genomics_heatmap(v$E[sig,], col_groups = col_group_names, row_k = 2,
            x = pData(T47D_eset)$title, y = fData(T47D_eset)$symbol[sig], 
            name = "RNA-seq", cbg = colorbar_grid(x_spacing = 0.2, nrow = 3),
            colorscale = diverging_colorscale(c(scales::muted("blue"),"white",scales::muted("red")))) 

rna_hm
```


  
#Get T47D Progesterone  
```{r T47D_tss_cvg}

tss_cvg_mats <- lapply(T47D, function(x){
  tmp <- make_coverage_matrix(inputs = fi$File.bam[x],
                                  ranges = tss_ranges[sig],
                                  binsize = 100,
                                  up = 5000,
                                  down = 5000)
  out <- Reduce("+", tmp) / length(tmp)
  return(out)
})

promoter_ranges = GenomicRanges::promoters(tss_ranges,
                                                 up = 5000,
                                                 down = 5000)

er_promoter_counts <- simplify2array(
  BiocParallel::bplapply(fi$File.bam[unlist(T47D[1:3])],
                         bamsignals::bamCount,
                         promoter_ranges[sig])) %>% 
  scale(center = FALSE, scale = sf[unlist(T47D[1:3])])

pr_promoter_counts <- simplify2array(
  BiocParallel::bplapply(fi$File.bam[unlist(T47D[4:6])],
                         bamsignals::bamCount,
                         promoter_ranges[sig])) %>% 
  scale(center = FALSE, scale = sf[unlist(T47D[4:6])])

p300_promoter_counts <- simplify2array(
  BiocParallel::bplapply(fi$File.bam[unlist(T47D[7:9])],
                         bamsignals::bamCount,
                         promoter_ranges[sig])) %>% 
  scale(center = FALSE, scale = sf[unlist(T47D[7:9])])


colnames(er_promoter_counts) <- unlist(fi$source_name[unlist(T47D[1:3])], use.names= FALSE)
er_chip_condition <- unlist(fi$condition[unlist(T47D[1:3])], use.names= FALSE)


colnames(pr_promoter_counts) <- unlist(fi$source_name[unlist(T47D[4:6])], use.names= FALSE)
pr_chip_condition <- unlist(fi$condition[unlist(T47D[4:6])], use.names= FALSE)

colnames(p300_promoter_counts) <- unlist(fi$source_name[unlist(T47D[7:9])], use.names= FALSE)
p300_chip_condition <- unlist(fi$condition[unlist(T47D[7:9])], use.names= FALSE)

#COLLAPSE CHIP

er_promoter_counts_collapsed <- sapply(unique(er_chip_condition),
                                      function(x){
                                        tmp <- which(er_chip_condition == x)
                                        apply(er_promoter_counts[,tmp], 1, sum, na.rm = TRUE)
                                      })

pr_promoter_counts_collapsed <- sapply(unique(pr_chip_condition),
                                      function(x){
                                        tmp <- which(pr_chip_condition == x)
                                        apply(pr_promoter_counts[,tmp], 1, sum, na.rm = TRUE)
                                      })

p300_promoter_counts_collapsed <- sapply(unique(p300_chip_condition),
                                      function(x){
                                        tmp <- which(p300_chip_condition == x)
                                        apply(p300_promoter_counts[,tmp], 1, sum, na.rm = TRUE)
                                      })

er_promoter_counts_diff <- log2(er_promoter_counts_collapsed[,2:3] / er_promoter_counts_collapsed[,1])
pr_promoter_counts_diff <- log2(pr_promoter_counts_collapsed[,2:3] / pr_promoter_counts_collapsed[,1])
p300_promoter_counts_diff <- log2(p300_promoter_counts_collapsed[,2:3] / p300_promoter_counts_collapsed[,1])
er_untreated <- er_promoter_counts_collapsed[,1]
pr_untreated <- pr_promoter_counts_collapsed[,1]
p300_untreated <- p300_promoter_counts_collapsed[,1]
untreated_range <- range(c(er_untreated, pr_untreated, p300_untreated))
diff_range <- range(c(er_promoter_counts_diff, pr_promoter_counts_diff, p300_promoter_counts_diff))


rna_chip_collapsed_hm <- rna_hm %>% 
  add_genomics_heatmap(er_promoter_counts_diff,
                       col_groups = c("P4","R5020"),
                       col_groups_name = "Treatment",
                       col_groups_palette = "Paired",
                       colorscale = diverging_colorscale("PiYG"),
                       scale = "none",
                       name = "ER ChIP<br>vs untreated",
                       size = 0.2, buffer = 0.1,
                       x_title = "ER",
                       zmin = diff_range[1],
                       zmax = diff_range[2]) %>%
  add_row_signal(er_untreated, 
                 name = "Untreated", 
                 x_title= 'ER<br> Untreated', 
                 colorscale = continuous_colorscale("Greys"),
                 zmin = untreated_range[1], 
                 zmax = untreated_range[2]) %>% 
  add_genomics_heatmap(pr_promoter_counts_diff,
                       col_groups = c("Progesterone","R5020"),
                       col_groups_palette = "Paired",
                       colorscale = diverging_colorscale("PiYG"),
                       scale = "none",
                       show_col_groups_colorbar = FALSE,
                       show_color = FALSE,
                       name = "PR ChIP<br>vs untreated",
                       size = 0.2, 
                       buffer = 0.1,
                       x_title = "PR",
                       zmin = diff_range[1],
                       zmax = diff_range[2]) %>%
  add_row_signal(pr_untreated, 
                 name = "PR ChIP <br> Untreated", 
                 x_title = 'PR<br> Untreated', 
                 colorscale = continuous_colorscale("Greys"), 
                 show_colorbar = FALSE,
                 zmin = untreated_range[1], 
                 zmax = untreated_range[2]) %>%
  add_genomics_heatmap(p300_promoter_counts_diff,
                       col_groups = c("Progesterone","R5020"),
                       col_groups_palette = "Paired",
                       show_col_groups_colorbar = FALSE,
                       show_color = FALSE,
                       colorscale = diverging_colorscale("PiYG"),
                       scale = "none",
                       name = "p300 ChIP<br>vs untreated",
                       size = 0.2, 
                       buffer = 0.1,
                       x_title = "p300",
                       zmin = diff_range[1],
                       zmax = diff_range[2]) %>%
  add_row_signal(p300_untreated, 
                 name="p300 ChIP <br> Untreated", 
                 x_title = 'p300<br> Untreated',
                 colorscale = continuous_colorscale("Greys"), 
                 show_colorbar = FALSE,
                 zmin = untreated_range[1], 
                 zmax = untreated_range[2]) %>%
  plot_iHeatmap() %>% layout(margin = list(l=0, t=0))

```


```{r Match transcripts with symbol, echo = FALSE}
tr <- transcriptsBy(TxDb.Hsapiens.BioMart.igis, by = "gene") %>% unlist()
tr$entrez <- stringr::str_extract(names(tr),"(?<=:)(.*)")
tr$symbol <- annotate::getSYMBOL(tr$entrez, data='org.Hs.eg')
names(tr) = NULL
trdf <- as.data.frame(tr)

truniq <- trdf %>% group_by(entrez) %>% summarize(chr = first(seqnames),
                                                  start = min(start),
                                                  end = min(end),
                                                  strand = first(strand),
                                                  symbol = first(symbol))
tx_data <- load_tx_data(TxDb.Hsapiens.BioMart.igis)
```

```{r Gene Loci, echo=FALSE}
symbol <- "ZNF579"
sym_idx <- which(truniq$symbol == symbol)
symbol_range <- get_view_range(chr = truniq$chr[sym_idx], start = truniq$start[sym_idx],
                               end = truniq$end[sym_idx])
symbol_range <- extend_grange(symbol_range, 50000)
range_cvg <- lapply(seq_along(T47D), function(i){
  x <- T47D[[i]]
  n <- paste0(rep(names(T47D)[[i]], length(x)), seq_along(x))
  tmp <- get_coverage_in_range(bwList = fi$File.bw[x], target_range = symbol_range,
                               names = n, cvg_scaling = sf[x])})
range_cvg <- Reduce(c, range_cvg)

```

```{r}
plot_browserly_tracks(symbol_range, tx_data, range_cvg, binsize = 100)
```


```{r}
cvg_subset <- c('PR_ab_Pr_treat1', 'PR_ab5')
plot_browserly_tracks(symbol_range, tx_data, range_cvg[cvg_subset], binsize=100)
```