---
title: "trackview_testing"
author: "Justin Finkle"
date: "7/5/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(chipVis)
library(gChipseq)
library(rtracklayer)
library(TxDb.Hsapiens.BioMart.igis)
library(SNPlocs.Hsapiens.dbSNP141.GRCh38)
```

## R Markdown

This is intended to demonstrate how the chipVis package displays local views.
```{r Load Data}
# Specify parameters
info_dir <- "~/CARROLL_PR/data/"
s_file <- paste(info_dir, "samples.txt", sep='')
p_file <- paste(info_dir, "pairs.txt", sep='')
fi <- getChipFileInfo(s_file, p_file)
symbol <- "NRIP1"
org <- "human"
genome <- "GRCh38"
```

```{r Load BigWig}
extension <- 50000
# Create the range for viewing. Can either specify a symbol or a range
range.gene <- getViewRange(symbol, org, genome)
# range.gene <- getViewRange(chr = 1, start=6000, end=6050)

# Extend the region if desired
range.gene <- extendGRange(range.gene, extension)

# Get the exon list
full.exon <- chipVis::igisExonlist(symbol, org=org, genome=genome)

# Get the coverage
input <- fi[2, 'Sample.Control']
input_idx <- which(fi$Sample.Name==input)
idx <- c(2:5, input_idx)
shortened_names <- c(fi[idx[1:length(idx)-1], "Sample.Name"], "Input")

# Get scaling factor
sf <- getScalingFactor(fi[idx,])/1000000
sf <- sf[2,]

cvg <- getCoverageInRange(fi[idx, "File.bw"], range.gene, shortened_names, sf)
```

```{r Basic Track}
track_info <- chipVis::plotGeneCoverage(grl = cvg, range.gene, sync = TRUE, genome=genome, 
                          symbol=symbol, dat.exon=full.exon)
```