---
title: "trackview_testing"
author: "Justin Finkle"
date: "7/5/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(gChipseq)
library(chipVis)
library(rtracklayer)
library(TxDb.Mmusculus.BioMart.igis3.0)
```

## R Markdown

This is intended to demonstrate how the chipVis package displays local views.

First we load some data to use for visualization
```{r Load Data}
# Specify parameters
info_dir <- "~/workspace/BAP1/"
s_file <- paste(info_dir, "BAP1_june_2016_samples.txt", sep='')
p_file <- paste(info_dir, "BAP1_june_2016_pairs.txt", sep='')
fi <- get_chip_file_info(s_file, p_file)
symbol <- "RNF2"
org <- "mouse"
genome <- "GRCm38"
tx_data <- load_tx_data(TxDb.Mmusculus.BioMart.igis)
txdb <- TxDb.Mmusculus.BioMart.igis
```


Next we need to specify the range that will be shown in the track. This can be done
in one of two ways:
  1. Get the range of a gene by supplying the gene symbol. This works by finding the transcripts
     that match the gene symbol and then calculating the corresponding range.
  2. Set the range manually by specifying a chromosome, start, and end position
After this the range can be extended, transcript anntotations can be obtained, and the
relevant coverage from the data samples can be pulled with the appropriate scaling factors.
```{r Load BigWig}
extension <- 150000
# Create the range for viewing. Can either specify a symbol or a range
# view_range <- get_view_range(tx_data = tx_data, db_object = txdb, symbol = symbol, 
                             # keytype = "SYMBOL")
view_range <- get_view_range(chr = 1, start=159758000, end=159758000)

# Extend the region if desired
view_range <- extend_grange(view_range, extension)

# Get the exon list
exon_data <- get_tx_annotation(db_object = txdb, range = view_range, 
                               tx_data = tx_data, no_introns=TRUE)

# Get the coverage
input <- fi[2, 'Sample.Control']
input_idx <- which(fi$Sample.Name==input)
idx <- c(9,10,15,16,18:23)
shortened_names <- c(fi[idx[1:length(idx)-1], "Genotype"], "Input")

# Get scaling factor
sf <- get_scaling_factor(fi[idx,])/1000000
sf <- sf[2,]

cvg <- get_coverage_in_range(fi[idx, "File.bw"], view_range, shortened_names, sf)
```

If available, the user can also load SNP data and TSS data. TSS annotations may be useful
if there are a lot of transcripts that will be collapsed to save vertical space.
```{r Extra annotation Information}
snps <- readRDS("~/workspace/ExternalData/SNP/gwas_catalog_v1.0.1-associations_e84_r2016-07-10.rds")
tss_db <- tssFromTxdb(txdb)
```

Now we can plot the track. See the "Trackview Options Demonstration" vignette for a more
detailed layout of all the options that can be used from GViz.
```{r Basic Track}
track_info <- plot_track_view(cvg_list = cvg, target_range = view_range, 
                              genome=genome, exon_data = exon_data, 
                              snp_gr = snps, tss_gr = tss_db)
```

Additionally, the plotting function can be wrapped into a simpler function that can be called by
multiple different functions. This provides a central location for modifying the attributes
of the trackview if it will be used in several places. The function `make_track_function`
takes all the arguments of `plot_track_view`, in addition to arguments for setup (e.g. the
coverage file names, scaling factor) so that it can be abstracted away.

```{r Function Abstraction}
cvg_files <- fi[idx, 'File.bw']
names(cvg_files) <- fi[idx, 'Genotype']
ptv <- make_track_function(target_range = view_range, cvg_files = cvg_files,
                           genome = genome, db_object = txdb, tx_data = tx_data,
                           cvg_scaling = sf, plot_kwargs = list(sizes=c(0.2, 0.3, 0.3)), grtrack_kwargs = list(stacking='squish'))
ptv(view_range)
```